include/nbl/builtin/hlsl/bxdf/base/cook_torrance_base.hlsl

```diff
@@ -0,0 +1,378 @@
+    static bool __call(NBL_CONST_REF_ARG(F) fresnel, NBL_CONST_REF_ARG(MicrofacetCache) cache)
+    {
+        return cache.isValid(fresnel.getRefractionOrientedEta());
+    }
+};
+
+template<class F, bool IsBSDF NBL_STRUCT_CONSTRAINABLE>
+struct getOrientedFresnel;
+
+template<class F>
+struct getOrientedFresnel<F, false>
+{
+    static F __call(NBL_CONST_REF_ARG(F) fresnel, typename F::scalar_type NdotV)
+    {
+        // expect conductor fresnel
+        return fresnel;
+    }
+};
+
+template<class F>
+NBL_PARTIAL_REQ_TOP(fresnel::TwoSidedFresnel<F>)
+struct getOrientedFresnel<F, true NBL_PARTIAL_REQ_BOT(fresnel::TwoSidedFresnel<F>) >
+{
+    using scalar_type = typename F::scalar_type;
+
+    static F __call(NBL_CONST_REF_ARG(F) fresnel, scalar_type NdotV)
+    {
+        return fresnel.getReorientedFresnel(NdotV);
+    }
+};
+}
+
+// N (NDF), F (fresnel)
+template<class Config, class N, class F NBL_PRIMARY_REQUIRES(config_concepts::MicrofacetConfiguration<Config> && ndf::NDF<N> && fresnel::Fresnel<F>)
+struct SCookTorrance
+{
+    MICROFACET_BXDF_CONFIG_TYPE_ALIASES(Config);
+
+    using this_t = SCookTorrance<Config, N, F>;
+    using quant_type = typename N::quant_type;
+    using ndf_type = N;
+    using fresnel_type = F;
+
+    NBL_CONSTEXPR_STATIC_INLINE bool IsAnisotropic = ndf_type::IsAnisotropic;
+    NBL_CONSTEXPR_STATIC_INLINE bool IsBSDF = ndf_type::NDFSurfaceType != ndf::MTT_REFLECT;
+
+    template<class Interaction, class MicrofacetCache>
+    spectral_type __eval(NBL_CONST_REF_ARG(sample_type) _sample, NBL_CONST_REF_ARG(Interaction) interaction, NBL_CONST_REF_ARG(MicrofacetCache) cache)
+    {
+        bool valid;
+        fresnel_type _f = fresnel;
+        NBL_IF_CONSTEXPR(IsBSDF)
+        {
+            _f = impl::getOrientedFresnel<fresnel_type, IsBSDF>::__call(fresnel, interaction.getNdotV());
+            valid = impl::check_TIR_helper<fresnel_type, IsBSDF>::template __call<MicrofacetCache>(_f, cache);
+        }
+        else
+            valid = _sample.getNdotL() > numeric_limits<scalar_type>::min && interaction.getNdotV() > numeric_limits<scalar_type>::min;
+        if (!valid)
+            return hlsl::promote<spectral_type>(0.0);
+
+        using quant_query_type = typename ndf_type::quant_query_type;
+        scalar_type dummy;
+        quant_query_type qq = ndf.template createQuantQuery<MicrofacetCache>(cache, dummy);
+
+        quant_type D = ndf.template D<sample_type, Interaction, MicrofacetCache>(qq, _sample, interaction, cache);
+        scalar_type DG = D.projectedLightMeasure;
+        if (D.microfacetMeasure < bit_cast<scalar_type>(numeric_limits<scalar_type>::infinity))
+        {
+            using g2g1_query_type = typename ndf_type::g2g1_query_type;
+            g2g1_query_type gq = ndf.template createG2G1Query<sample_type, Interaction>(_sample, interaction);
+            DG *= ndf.template correlated<sample_type, Interaction>(gq, _sample, interaction);
+        }
```

Comment:
hmm actually since Eval is not Quotient, it should return 0 whenever PDF would have been INF, and therefore whenever Eval is inf, you can just return 0

you could even add an `NBL_REF_ARG(bool)` to the `D` signature for a boolean side-channel of `isInf` because it pays for the NDF to set that boolean as soon as possible (before a chain or multiplies and divides, compiler won't hoiust your INF checks before divisions by measure factors, etc.)